Add some documentation around templates in config policies (#261)
open-cluster-management-io/open-cluster-management-io.github.io: content/en/getting-started/integration/policy-framework.md

and the resource definition generated from the resolved template must be a valid YAML. 
(See the [Golang documentation about package templates](https://golang.org/pkg/text/template/) 
for more information.) Any errors in template validation appear as policy violations. 
When you use a custom template function, the values are replaced at runtime.

Template functions, such as resource-specific and generic `lookup` template functions, are 
available for referencing Kubernetes resources on the cluster. The resource-specific functions 
are used for convenience and makes content of the resources more accessible. In addition to these 
functions, utility functions like `base64encode`, `base64decode`, `indent`, `autoindent`, `toInt`, 
`toBool`, and more are also available.

To conform templates to YAML syntax, templates must be set in the policy resource as 
strings using quotes or a block character (`|` or `>`). This causes the resolved template value 
to also be a string. To override this, consider using `toInt` or `toBool` as the final function 
in the template to initiate further processing that forces the value to be interpreted 
as an integer, or boolean.

> **Note**: If the string value is more than 80 characters, the block character `|` 

Reviewer: This is no longer true in OCM 0.6.0. This can be removed.

- Override a single template by wrapping the template in additional braces. For example, the 
  template `{{ template content }}` would become `{{ '{{ template content }}' }}`.
- Override all templates in a `ConfigurationPolicy` by adding the 
  `policy.open-cluster-management.io/disable-templates: "true"` annotation in the 
  `ConfigurationPolicy` section of your `Policy`. Template processing will be bypassed for 
  that `ConfigurationPolicy`.

#### Hub cluster templates

Hub cluster templates are used to define configuration policies that are dynamically 
customized to the target cluster. This reduces the need to create separate policies 
for each target cluster or hardcode configuration values in the policy definitions. 

Hub cluster templates are based on Golang text template specifications, and the `{{hub … hub}}` 
delimiter indicates a hub cluster template in a configuration policy.

A configuration policy definition can contain both hub cluster and managed cluster 
templates. Hub cluster templates are processed first on the hub cluster, then the policy 
definition with resolved hub cluster templates is propagated to the target clusters. 

> On the managed cluster, the `ConfigurationPolicyController` processes any managed cluster 

Reviewer: ```suggestion
On the managed cluster, the Configuration Policy controller processes any managed cluster 
```

    - policy-subscriptions
```

### Policy Templates

> Configuration policies support the inclusion of Golang text templates in the object definitions. These templates are resolved at runtime either on the hub cluster or the target managed cluster using configurations related to that cluster. This gives you the ability to define configuration policies with dynamic content, and inform or enforce Kubernetes resources that are customized to the target cluster.

Reviewer 2: Please reflow this text so that it matches the line breaks in the surrounding text.

> Configuration policies support the inclusion of Golang text templates in the object definitions. These templates are resolved at runtime either on the hub cluster or the target managed cluster using configurations related to that cluster. This gives you the ability to define configuration policies with dynamic content, and inform or enforce Kubernetes resources that are customized to the target cluster.

Reviewer 2: ```suggestion
Configuration policies support the inclusion of Golang text templates in the object definitions. These templates are resolved at runtime either on the hub cluster or the target managed cluster using configurations related to that cluster. This gives you the ability to define configuration policies with dynamic content and to inform or enforce Kubernetes resources that are customized to the target cluster.
```

> * The template syntax must be conformed to the Golang template language specification, and the resource definition generated from the resolved template must be a valid YAML. See the Golang documentation about https://golang.org/pkg/text/template/[Package templates] for more information. Any errors in template validation are recognized as policy violations. When you use a custom template function, the values are replaced at runtime.

Reviewer 2: ```suggestion
* The template syntax must follow the Golang template language specification, and the resource definition generated from the resolved template must be a valid YAML. (See the [Golang documentation about package templates](https://pkg.go.dev/text/template) for more information.) Any errors in template validation appear as policy violations. When you use a custom template function, the values are replaced at runtime.
```

> Template functions, such as resource-specific and generic, `lookup` template functions, are available for referencing Kubernetes resources on the cluster. The resource-specific functions are used for convenience and makes content of the resources more accessible. If you use the generic function, `lookup`, which is more advance, it is best to be familiar with the YAML structure of the resource that is being looked up. In addition to these functions, utility functions like `base64encode`, `base64decode`, `indent`, `autoindent`, `toInt`, `toBool` etc. are also available.

Reviewer 2: ```suggestion
Template functions, such as resource-specific and generic `lookup` template functions, are available for referencing Kubernetes resources on the cluster. The resource-specific functions are used for convenience and makes content of the resources more accessible. In addition to these functions, utility functions like `base64encode`, `base64decode`, `indent`, `autoindent`, `toInt`, `toBool`, and more are also available.
```

> To conform templates into YAML syntax, templates must be set in the policy resource as strings using quotes or a block character (`|` or `>`). This causes the resolved template value to also be a string. To override this, consider using `toInt` or `toBool` as the final function in the template to initiate further processing that forces the value to be interpreted as an integer, or boolean.

Reviewer 2: ```suggestion
To conform templates to YAML syntax, templates must be set in the policy resource as strings using quotes or a block character (`|` or `>`). This causes the resolved template value to also be a string. To override this, consider using `toInt` or `toBool` as the final function in the template to initiate further processing that forces the value to be interpreted as an integer, or boolean.
```

> *Note*: If the string value is more than 80 characters, this block character, `|` needs to be used to avoid YAML parsing errors.

Reviewer 2: ```suggestion
**Note**: If the string value is more than 80 characters, the block character `|` must be used to avoid YAML parsing errors.
```

> To bypass template processing for your hub cluster, you must change `{{ template content }}` to `{{ '{{ template content }}' }}`.
> 
> Alternatively, you can add the following annotation in the `ConfigurationPolicy` section of your `Policy`: `policy.open-cluster-management.io/disable-templates: "true"`. When this annotation is included, the previous workaround is not necessary. Template processing is bypassed for the `ConfigurationPolicy`.

Reviewer 2: ```suggestion
To bypass template processing for your hub cluster, you can either:
- Override a single template by wrapping the template in additional braces. For example, the template `{{ template content }}` would become `{{ '{{ template content }}' }}`.
- Override all templates in a `ConfigurationPolicy` by adding the `policy.open-cluster-management.io/disable-templates: "true"` annotation in the `ConfigurationPolicy` section of your `Policy`. Template processing will be bypassed for that `ConfigurationPolicy`.
```

| `fromClusterClaim` | The `fromClusterClaim` function returns the value of the `Spec.Value` in the `ClusterClaim` resource. | `platform: '{{ fromClusterClaim "platform.open-cluster-management.io" }}'` |
| `lookup` | The `lookup` function returns the Kubernetes resource as a JSON compatible map. Note that if the requested resource does not exist, an empty map is returned. | `metrics-url: \|`<br />`http://{{ (lookup "v1" "Service" "default" "metrics").spec.clusterIP }}:8080` |
| `base64enc` | The `base64enc` function returns a `base64` encoded value of the input `data string`. | `USER_NAME: '{{ fromConfigMap "default" "myconfigmap" "admin-user" \| base64enc }}'` |
| `base64dec` | The `base64dec` function returns a `base64` decoded value of the input `enc-data string`. | `app-name: \|`<br />`"{{ ( lookup "v1"  "Secret" "testns" "mytestsecret") .data.appname ) \| base64dec }}"` |
| `indent` | The `indent` function returns the padded `data string`. | `Ca-cert:  \|`<br />`{{ ( index ( lookup "v1" "Secret" "default" "mycert-tls"  ).data  "ca.pem"  ) \|  base64dec \| indent 4  }}` |

> | `autoindent` | The `autoindent` function acts like the `indent` function that automatically determines the number of leading spaces based on the number of spaces before the template. | `Ca-cert:  \|`<br />`{{ ( index ( lookup "v1" "Secret" "default" "mycert-tls"  ).data  "ca.pem"  ) \|  base64dec \| autoindent }}` |

Author: Should be ok, getting ca certs to the right namespaces is valuable here.

> | `toInt` | The `toInt` function casts and returns the integer value of the input value. Also, when this is the last function in the template, there is further processing of the source content. This is to ensure that the value is interpreted as an integer by the YAML. | `vlanid:  \|`<br />`{{ (fromConfigMap "site-config" "site1" "vlan")  \| toInt }}` |

Reviewer 2: ```suggestion
| `toInt` | Returns the integer value of the input string and ensures that the value is interpreted as an integer in the YAML. | `vlanid:  \|`<br />`{{ (fromConfigMap "site-config" "site1" "vlan")  \| toInt }}` |
```

> | `toBool` | The `toBool` function converts the input string into a boolean, and returns the boolean. Also, when this is the last function in the template, there is further processing of the source content. This is to ensure that the value is interpreted as a boolean by the YAML. | `enabled:  \|`<br />`{{ (fromConfigMap "site-config" "site1" "enabled")  \| toBool }}` |

Reviewer 2: ```suggestion
| `toBool` | Returns the boolean value of the input string and ensures that the value is interpreted as a boolean in the YAML. | `enabled:  \|`<br />`{{ (fromConfigMap "site-config" "site1" "enabled")  \| toBool }}` |
```

> | `protect` | The `protect` function enables you to encrypt a string in a hub cluster policy template. It is automatically decrypted on the managed cluster when the policy is evaluated. On the replicated policy in the managed cluster namespace, the value might resemble the following syntax: `$ocm_encrypted:okrrBqt72oI+3WT/0vxeI3vGa+wpLD7Z0ZxFMLvL204=` | `enabled: \|`<br />`{{hub "(lookup "v1" "Secret" "default" "my-hub-secret").data.message \| protect hub}}` |

Reviewer 2: ```suggestion
| `protect` | Encrypts the input string. It is decrypted when the policy is evaluated. On the replicated policy in the managed cluster namespace, the resulting value might resemble the following: `$ocm_encrypted:okrrBqt72oI+3WT/0vxeI3vGa+wpLD7Z0ZxFMLvL204=` | `enabled: \|`<br />`{{hub "(lookup "v1" "Secret" "default" "my-hub-secret").data.message \| protect hub}}` |
```

Reviewer 2: I don't think this is restricted to hub templates? Also, I think using `lookup "Secret"` with `protect` is somewhat of an antipattern since I think `fromSecret` should do protection, so probably use a different kind as above?

> #### Hub cluster templates

Reviewer 2: I wonder if much of this section can be consolidated/simplified and moved up to the top "Policy Templates" section.

Policies are processed on the hub cluster only upon creation or after an update. Therefore, hub cluster templates are only resolved to the data in the referenced resources upon policy creation or update. Any changes to the referenced resources are not automatically synced to the policies. 

A special annotation, `policy.open-cluster-management.io/trigger-update` can be used to indicate changes to the data referenced by the templates. Any change to the special annotation value initiates template processing, and the latest contents of the referenced resource are read and updated in the policy definition that is the propagator for processing on managed clusters. A typical way to use this annotation is to increment the value by one each time.

> #### Additional hub template encryption details
> 
> The encryption algorithm used is AES-CBC using 256-bit keys. Each encryption key is unique per managed cluster and is automatically rotated every 30 days.
> 
> This ensures that your decrypted value is to never be stored in the policy on the managed cluster.
> 
> To force an immediate rotation, delete the `policy.open-cluster-management.io/last-rotated` annotation on the `policy-encryption-key` Secret in the managed cluster namespace on the hub cluster. Policies are then reprocessed to use the new encryption key. 

Reviewer 2: ```suggestion
#### Template encryption details

The encryption algorithm uses AES-CBC with 256-bit keys. Each encryption key is unique per managed cluster and is automatically rotated every 30 days. This ensures that your decrypted value is never stored in the policy on the managed cluster.

To force an immediate encryption key rotation, delete the `policy.open-cluster-management.io/last-rotated` annotation on the `policy-encryption-key` Secret in the managed cluster namespace on the hub cluster. Policies are then reprocessed to use the new encryption key. 
```
